### ISSUE TITLE
Instability when resuming trains

### ISSUE THREAD

**Issue by angusturner** [lucidrains/lion-pytorch#13]
Hi, I have been testing this out on some diffusion models I am training.

Convergence seems decent (somewhat faster than AdamW, using 1/10th the learning rate, 10x weight decay).
However, I recently paused a few experiments and tried to resume, and the loss explodes immediately. I do not
face this issue when resuming AdamW trains.

I have also found it necessary to use an LR warm-up period in my trains (even with the 1/10th loss), which again,
is not required in AdamW. I'll try to do a bit more digging to see if I can track down the source of instability - however for experiment resuming, surely if I load the optimizer state correctly things should resume as expected?

My only thought is whether something could be going wrong with saving of EMA / moving average statistics? If I get a chance to dig into this more I'll let you know what I find. (Possibly I am doing something wrong).

**Comment by xiangning-chen** [comment]
Hi, thanks for testing the Lion optimizer.
Just wondering what's the betas for AdamW and Lion, are they both as default, i.e., (0.9, 0.999) for AdamW, and (0.9, 0.99) for Lion?

**Comment by angusturner** [comment]
No worries! It's a really cool idea, so it will be nice if it can consistently improve on Adam! It is too early to say in my own experiments. For Lion I am using default (0.9, 0.99). For AdamW I have had successful runs with the defaults (0.9, 0.999).  and also using (0.9, 0.99). Still tuning, so not sure what the optimum values are.

**Comment by xiangning-chen** [comment]
Sounds good. Another betas setting (0.95, 0.98) can help with the training stability.

**Comment by lucidrains** [comment]
@xiangning-chen have you run into this stability issue yourself?

**Comment by xiangning-chen** [comment]
Not on diffusion model, when I encountered instability I just lower the learning rate. On language modeling, I found that lowering the beta2 improves stability for both AdamW and Lion.

**Comment by clementpoiret** [comment]
Betas (0.95, 0.98) considerably lowered my instabilities, thanks for the tip @xiangning-chen !
I still have heavy instabilities as soon as I unfreeze some layers during fine-tuning sessions, so I just try to lower the LR.

**Comment by lucidrains** [comment]
@clementpoiret nice! 🙏

**Comment by xiangning-chen** [comment]
@clementpoiret Thanks for the update! For fine-tuning, are you referring to using Lion to fine-tune an AdamW trained model?

**Comment by clementpoiret** [comment]
@xiangning-chen, Yup, it's a pretrained EfficientNet from timm. I replaced the classifier with my own MLP.
For 1/4th of the epochs I train only my MLP, then I gradually unfreeze the last two blocks based on an EarlyStopping strategy. You can see that everything explodes, and it does not happen using AdamW, or when I only train the MLP without fine-tuning layers. I also use gradient clipping.
![image](https://user-images.githubusercontent.com/10899984/223055119-a994cef5-231a-4989-9b89-e50c3a06b924.png)

**Comment by xiangning-chen** [comment]
Did you load the AdamW optimizer status from pre-training?

**Comment by clementpoiret** [comment]
Not at all, I just loaded the weights from timm. (Please note that the strange double descent at the start is normal, I also have it using other optimizers.)

**Comment by lucidrains** [comment]
@xiangning-chen do you have any experiments showing that loading adam momentum into lion for fine tuning is better? happy to add that feature, provided it isn't just a hunch you have

**Comment by lucidrains** [comment]
oh, actually, loading adam optimizer state works fine as is, ok no worries

**Comment by xiangning-chen** [comment]
> do you have any experiments showing that loading adam momentum into lion for fine tuning is better? happy to add that feature, provided it isn't just a hunch you have

Oh I meant when using both Adam for pre-training and fine-tuning, loading the 1st and 2nd moments are helpful. Never tried loading the Adam momentum into Lion as their EMA parameters are different.

**Comment by flymark2010** [comment]
> Hi, I have been testing this out on some diffusion models I am training.
> 
> Convergence seems decent (somewhat faster than AdamW, using 1/10th the learning rate, 10x weight decay). However, I recently paused a few experiments and tried to resume, and the loss explodes immediately. I do not face this issue when resuming AdamW trains.
> 
> I have also found it necessary to use an LR warm-up period in my trains (even with the 1/10th loss), which again, is not required in AdamW. I'll try to do a bit more digging to see if I can track down the source of instability - however for experiment resuming, surely if I load the optimizer state correctly things should resume as expected?
> 
> My only thought is whether something could be going wrong with saving of EMA / moving average statistics? If I get a chance to dig into this more I'll let you know what I find. (Possibly I am doing something wrong).

I got the same problem that  loss explodes immediately when resuming checkpoints when `use_triton=True` in NLP task.  Then setting `use_triton=False`, the loss becomes normal. 

I'm not familiar with triton, but I guess using [triton.language.load](https://github.com/lucidrains/lion-pytorch/blob/ee26e32b1cf06fe4afe043883e668b0ed87a0d8a/lion_pytorch/triton.py#L42) and [triton.language.store](https://github.com/lucidrains/lion-pytorch/blob/ee26e32b1cf06fe4afe043883e668b0ed87a0d8a/lion_pytorch/triton.py#L71) may cache some states, which will be cleaned after resuming.

**Comment by mitchellnw** [comment]
I think this is because of https://github.com/lucidrains/lion-pytorch/issues/24#issuecomment-1539290226

**Comment by ipoletaev** [comment]
Shall we close this one as a duplicate of https://github.com/lucidrains/lion-pytorch/issues/20 ?

**Comment by lucidrains** [comment]
@ipoletaev ah yes, Mitchell filled me in on this issue through email this morning

do you want to see if https://github.com/lucidrains/lion-pytorch/commit/6ab873a380b47ebc5ea6f68ea588606daebb8b85 resolves it? the more permanent solution would be to submit a PR to Triton to clone all tensors prior to auto-tuning

**Comment by ipoletaev** [comment]
IMO it makes sense to just remove autotune and keep it simple: where the user specifies the block size they need.

**Comment by lucidrains** [comment]
@ipoletaev yea true, i'll do that if this hack doesn't do the trick

**Comment by lucidrains** [comment]
@ipoletaev actually, you are right, this battle isn't worth fighting

https://github.com/lucidrains/lion-pytorch/commit/2226ec8aeee03e9fbbf561e50fbf114b9677d3e9
